packages/edit-site/src/components/add-new-template/add-custom-template-modal.js

```diff
@@ -0,0 +1,243 @@
+	SearchControl,
+	TextHighlight,
+	__experimentalText as Text,
+	__experimentalHeading as Heading,
+	__unstableComposite as Composite,
+	__unstableUseCompositeState as useCompositeState,
+	__unstableCompositeItem as CompositeItem,
+} from '@wordpress/components';
+import { useSelect } from '@wordpress/data';
+import { useDebounce } from '@wordpress/compose';
+import { store as coreStore } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import { mapToIHasNameAndId } from './utils';
+
+const EMPTY_ARRAY = [];
+const BASE_QUERY = {
+	order: 'asc',
+	_fields: 'id,title,slug,link',
+	context: 'view',
+};
+
+function SuggestionListItem( {
+	suggestion,
+	search,
+	onSelect,
+	entityForSuggestions,
+	composite,
+} ) {
+	const baseCssClass =
+		'edit-site-custom-template-modal__suggestions_list__list-item';
+	return (
+		<CompositeItem
+			role="option"
+			as={ Button }
+			{ ...composite }
+			className={ baseCssClass }
+			onClick={ () => {
+				const title = `${ entityForSuggestions.labels.singular }: ${ suggestion.name }`;
+				onSelect( {
+					title,
+					description: `Template for ${ title }`,
+					slug: `single-${ entityForSuggestions.slug }-${ suggestion.slug }`,
+				} );
+			} }
+		>
+			<span className={ `${ baseCssClass }__title` }>
+				<TextHighlight text={ suggestion.name } highlight={ search } />
+			</span>
+			{ suggestion.link && (
+				<span className={ `${ baseCssClass }__info` }>
+					{ suggestion.link }
+				</span>
+			) }
+		</CompositeItem>
+	);
+}
+
+function SuggestionList( { entityForSuggestions, onSelect } ) {
+	const composite = useCompositeState( { orientation: 'vertical' } );
+	const [ suggestions, setSuggestions ] = useState( EMPTY_ARRAY );
+	// We need to track two values, the search input's value(searchInputValue)
+	// and the one we want to debounce(search) and make REST API requests.
+	const [ searchInputValue, setSearchInputValue ] = useState( '' );
+	const [ search, setSearch ] = useState( '' );
+	const debouncedSearch = useDebounce( setSearch, 250 );
+	const { searchResults, searchHasResolved } = useSelect(
+		( select ) => {
+			const { getEntityRecords, hasFinishedResolution } =
+				select( coreStore );
+			const selectorArgs = [
+				entityForSuggestions.type,
+				entityForSuggestions.slug,
+				{
+					...BASE_QUERY,
+					search,
+					orderby: !! search ? 'relevance' : 'modified',
+					exclude: entityForSuggestions.postsToExclude,
+					per_page: !! search ? 20 : 10,
+				},
+			];
+			return {
+				searchResults: getEntityRecords( ...selectorArgs ),
+				searchHasResolved: hasFinishedResolution(
+					'getEntityRecords',
+					selectorArgs
+				),
+			};
+		},
+		[ search ]
+	);
+	useEffect( () => {
+		if ( search !== searchInputValue ) {
+			debouncedSearch( searchInputValue );
+		}
+	}, [ search, searchInputValue ] );
+	const entitiesInfo = useMemo( () => {
+		if ( ! searchResults?.length ) return EMPTY_ARRAY;
+		return mapToIHasNameAndId( searchResults, 'title.rendered' );
```

Comment:
Sorry about that and I can understand your confusion here. `mapToIHasNameAndId` will be used in the other follow ups for other templates and didn't comment about it because it was clear in my head, but I should.. The reason it will be used is that we will have different entities to handle(taxonomies, authors, etc..) and we need an interface to handle the results better.

This is related to your other comment about the naming of [defaultTemplates, extraTemplates](https://github.com/WordPress/gutenberg/pull/41189#discussion_r899228186). This PR had already at least other three PRs with preparatory work and it will be refactored and fine grained as we move to the other templates/entities.

Probably the `path` to get a `name` or the `fields` we want to show for an entity would be variables per entity, but we need to get one part in(in a good shape of course) and refactor/refine as we go..

Do you think I should add these right now? I can, but FYI I intend to work immediately on the other templates when we land this one.

Reply:
Following up separately is fine!